File: cmd/esbuild/stdio_protocol.go
// The JavaScript API communicates with the Go child process over stdin/stdout
// using this protocol. It's a very simple binary protocol that uses primitives
// and nested arrays and maps. You must send a response after receiving a
// request because the other end is blocking on the response coming back.

package main

import (
	"encoding/binary"
	"sort"
)

func readUint32(bytes []byte) (value uint32, leftOver []byte, ok bool) {
	if len(bytes) >= 4 {
		return binary.LittleEndian.Uint32(bytes), bytes[4:], true
	}

	return 0, bytes, false
}

func writeUint32(bytes []byte, value uint32) []byte {
	bytes = append(bytes, 0, 0, 0, 0)
	binary.LittleEndian.PutUint32(bytes[len(bytes)-4:], value)
	return bytes
}

func readLengthPrefixedSlice(bytes []byte) (slice []byte, leftOver []byte, ok bool) {
	if length, afterLength, ok := readUint32(bytes); ok && uint(len(afterLength)) >= uint(length) {
		return afterLength[:length], afterLength[length:], true
	}

	return []byte{}, bytes, false
}

type packet struct {
	id        uint32
	isRequest bool
	value     interface{}
}

func encodePacket(p packet) []byte {
	var visit func(interface{})
	var bytes []byte

	visit = func(value interface{}) {
		switch v := value.(type) {
		case nil:
			bytes = append(bytes, 0)

		case bool:
			n := uint8(0)
			if v {
				n = 1
			}
			bytes = append(bytes, 1, n)

		case int:
			bytes = append(bytes, 2)
			bytes = writeUint32(bytes, uint32(v))

		case string:
			bytes = append(bytes, 3)
			bytes = writeUint32(bytes, uint32(len(v)))
			bytes = append(bytes, v...)

		case []byte:
			bytes = append(bytes, 4)
			bytes = writeUint32(bytes, uint32(len(v)))
			bytes = append(bytes, v...)

		case []interface{}:
			bytes = append(bytes, 5)
			bytes = writeUint32(bytes, uint32(len(v)))
			for _, item := range v {
				visit(item)
			}

		case map[string]interface{}:
			// Sort keys for determinism
			keys := make([]string, 0, len(v))
			for k := range v {
				keys = append(keys, k)
			}
			sort.Strings(keys)
			bytes = append(bytes, 6)
			bytes = writeUint32(bytes, uint32(len(keys)))
			for _, k := range keys {
				bytes = writeUint32(bytes, uint32(len(k)))
				bytes = append(bytes, k...)
				visit(v[k])
			}

		default:
			panic("Invalid packet")
		}
	}

	bytes = writeUint32(bytes, 0) // Reserve space for the length
	if p.isRequest {
		bytes = writeUint32(bytes, p.id<<1)
	} else {
		bytes = writeUint32(bytes, (p.id<<1)|1)
	}
	visit(p.value)
	writeUint32(bytes[:0], uint32(len(bytes)-4)) // Patch the length in
	return bytes
}

func decodePacket(bytes []byte) (packet, bool) {
	var visit func() (interface{}, bool)

	visit = func() (interface{}, bool) {
		kind := bytes[0]
		bytes = bytes[1:]
		switch kind {
		case 0: // nil
			return nil, true

		case 1: // bool
			value := bytes[0]
			bytes = bytes[1:]
			return value != 0, true

		case 2: // int
			value, next, ok := readUint32(bytes)
			if !ok {
				return nil, false
			}
			bytes = next
			return int(value), true

		case 3: // string
			value, next, ok := readLengthPrefixedSlice(bytes)
			if !ok {
				return nil, false
			}
			bytes = next
			return string(value), true

		case 4: // []byte
			value, next, ok := readLengthPrefixedSlice(bytes)
			if !ok {
				return nil, false
			}
			bytes = next
			return value, true

		case 5: // []interface{}
			count, next, ok := readUint32(bytes)
			if !ok {
				return nil, false
			}
			bytes = next
			value := make([]interface{}, count)
			for i := 0; i < int(count); i++ {
				item, ok := visit()
				if !ok {
					return nil, false
				}
				value[i] = item
			}
			return value, true

		case 6: // map[string]interface{}
			count, next, ok := readUint32(bytes)
			if !ok {
				return nil, false
			}
			bytes = next
			value := make(map[string]interface{}, count)
			for i := 0; i < int(count); i++ {
				key, next, ok := readLengthPrefixedSlice(bytes)
				if !ok {
					return nil, false
				}
				bytes = next
				item, ok := visit()
				if !ok {
					return nil, false
				}
				value[string(key)] = item
			}
			return value, true

		default:
			panic("Invalid packet")
		}
	}

	id, bytes, ok := readUint32(bytes)
	if !ok {
		return packet{}, false
	}
	isRequest := (id & 1) == 0
	id >>= 1
	value, ok := visit()
	if !ok {
		return packet{}, false
	}
	if len(bytes) != 0 {
		return packet{}, false
	}
	return packet{id: id, isRequest: isRequest, value: value}, true
}

File: lib/shared/stdio_protocol.ts
// The JavaScript API communicates with the Go child process over stdin/stdout
// using this protocol. It's a very simple binary protocol that uses primitives
// and nested arrays and maps. It's basically JSON with UTF-8 encoding and an
// additional byte array primitive. You must send a response after receiving a
// request because the other end is blocking on the response coming back.

import * as types from "./types";

export interface BuildRequest {
  command: 'build';
  key: number;
  entries: [string, string][]; // Use an array instead of a map to preserve order
  flags: string[];
  write: boolean;
  stdinContents: string | null;
  stdinResolveDir: string | null;
  absWorkingDir: string;
  incremental: boolean;
  nodePaths: string[];
  plugins?: BuildPlugin[];
  serve?: ServeRequest;
}

export interface ServeRequest {
  serveID: number;
  port?: number;
  host?: string;
  servedir?: string;
}

export interface ServeResponse {
  port: number;
  host: string;
}

export interface ServeStopRequest {
  command: 'serve-stop';
  serveID: number;
}

export interface BuildPlugin {
  name: string;
  onResolve: { id: number, filter: string, namespace: string }[];
  onLoad: { id: number, filter: string, namespace: string }[];
}

export interface BuildResponse {
  errors: types.Message[];
  warnings: types.Message[];
  outputFiles: BuildOutputFile[];
  metafile: string;
  writeToStdout?: Uint8Array;
  rebuildID?: number;
  watchID?: number;
}

export interface BuildOutputFile {
  path: string;
  contents: Uint8Array;
  isEntryPoint?: boolean;
}

export interface PingRequest {
  command: 'ping';
}

export interface RebuildRequest {
  command: 'rebuild';
  rebuildID: number;
}

export interface RebuildDisposeRequest {
  command: 'rebuild-dispose';
  rebuildID: number;
}

export interface WatchStopRequest {
  command: 'watch-stop';
  watchID: number;
}

export interface OnRequestRequest {
  command: 'serve-request';
  serveID: number;
  args: types.ServeOnRequestArgs;
}

export interface OnWaitRequest {
  command: 'serve-wait';
  serveID: number;
  error: string | null;
}

export interface OnWatchRebuildRequest {
  command: 'watch-rebuild';
  watchID: number;
  args: types.BuildResult;
}

export interface TransformRequest {
  command: 'transform';
  flags: string[];
  input: string;
  inputFS: boolean;
}

export interface TransformResponse {
  errors: types.Message[];
  warnings: types.Message[];

  code: string;
  codeFS: boolean;

  map: string;
  mapFS: boolean;
}

export interface FormatMsgsRequest {
  command: 'format-msgs';
  messages: types.Message[];
  isWarning: boolean;
  color?: boolean;
  terminalWidth?: number;
}

export interface FormatMsgsResponse {
  messages: string[];
}

export interface OnStartRequest {
  command: 'start';
  key: number;
}

export interface OnStartResponse {
  errors?: types.PartialMessage[];
  warnings?: types.PartialMessage[];
}

export interface OnResolveRequest {
  command: 'resolve';
  key: number;
  ids: number[];
  path: string;
  importer: string;
  namespace: string;
  resolveDir: string;
  kind: types.ImportKind;
  pluginData: number;
}

export interface OnResolveResponse {
  id?: number;
  pluginName?: string;

  errors?: types.PartialMessage[];
  warnings?: types.PartialMessage[];

  path?: string;
  external?: boolean;
  sideEffects?: boolean;
  namespace?: string;
  pluginData?: number;

  watchFiles?: string[];
  watchDirs?: string[];
}

export interface OnLoadRequest {
  command: 'load';
  key: number;
  ids: number[];
  path: string;
  namespace: string;
  pluginData: number;
}

export interface OnLoadResponse {
  id?: number;
  pluginName?: string;

  errors?: types.PartialMessage[];
  warnings?: types.PartialMessage[];

  contents?: Uint8Array;
  resolveDir?: string;
  loader?: string;
  pluginData?: number;

  watchFiles?: string[];
  watchDirs?: string[];
}

////////////////////////////////////////////////////////////////////////////////

export interface Packet {
  id: number;
  isRequest: boolean;
  value: Value;
}

export type Value =
  | null
  | boolean
  | number
  | string
  | Uint8Array
  | Value[]
  | { [key: string]: Value }

export function encodePacket(packet: Packet): Uint8Array {
  let visit = (value: Value) => {
    if (value === null) {
      bb.write8(0);
    } else if (typeof value === 'boolean') {
      bb.write8(1);
      bb.write8(+value);
    } else if (typeof value === 'number') {
      bb.write8(2);
      bb.write32(value | 0);
    } else if (typeof value === 'string') {
      bb.write8(3);
      bb.write(encodeUTF8(value));
    } else if (value instanceof Uint8Array) {
      bb.write8(4);
      bb.write(value);
    } else if (value instanceof Array) {
      bb.write8(5);
      bb.write32(value.length);
      for (let item of value) {
        visit(item);
      }
    } else {
      let keys = Object.keys(value);
      bb.write8(6);
      bb.write32(keys.length);
      for (let key of keys) {
        bb.write(encodeUTF8(key));
        visit(value[key]);
      }
    }
  };

  let bb = new ByteBuffer;
  bb.write32(0); // Reserve space for the length
  bb.write32((packet.id << 1) | +!packet.isRequest);
  visit(packet.value);
  writeUInt32LE(bb.buf, bb.len - 4, 0); // Patch the length in
  return bb.buf.subarray(0, bb.len);
}

export function decodePacket(bytes: Uint8Array): Packet {
  let visit = (): Value => {
    switch (bb.read8()) {
      case 0: // null
        return null;
      case 1: // boolean
        return !!bb.read8();
      case 2: // number
        return bb.read32();
      case 3: // string
        return decodeUTF8(bb.read());
      case 4: // Uint8Array
        return bb.read();
      case 5: { // Value[]
        let count = bb.read32();
        let value: Value[] = [];
        for (let i = 0; i < count; i++) {
          value.push(visit());
        }
        return value;
      }
      case 6: { // { [key: string]: Value }
        let count = bb.read32();
        let value: { [key: string]: Value } = {};
        for (let i = 0; i < count; i++) {
          value[decodeUTF8(bb.read())] = visit();
        }
        return value;
      }
      default:
        throw new Error('Invalid packet');
    }
  };

  let bb = new ByteBuffer(bytes);
  let id = bb.read32();
  let isRequest = (id & 1) === 0;
  id >>>= 1;
  let value = visit();
  if (bb.ptr !== bytes.length) {
    throw new Error('Invalid packet');
  }
  return { id, isRequest, value };
}

class ByteBuffer {
  len = 0;
  ptr = 0;

  constructor(public buf = new Uint8Array(1024)) {
  }

  private _write(delta: number): number {
    if (this.len + delta > this.buf.length) {
      let clone = new Uint8Array((this.len + delta) * 2);
      clone.set(this.buf);
      this.buf = clone;
    }
    this.len += delta;
    return this.len - delta;
  }

  write8(value: number): void {
    let offset = this._write(1);
    this.buf[offset] = value;
  }

  write32(value: number): void {
    let offset = this._write(4);
    writeUInt32LE(this.buf, value, offset);
  }

  write(bytes: Uint8Array): void {
    let offset = this._write(4 + bytes.length);
    writeUInt32LE(this.buf, bytes.length, offset);
    this.buf.set(bytes, offset + 4);
  }

  private _read(delta: number): number {
    if (this.ptr + delta > this.buf.length) {
      throw new Error('Invalid packet');
    }
    this.ptr += delta;
    return this.ptr - delta;
  }

  read8(): number {
    return this.buf[this._read(1)];
  }

  read32(): number {
    return readUInt32LE(this.buf, this._read(4));
  }

  read(): Uint8Array {
    let length = this.read32();
    let bytes = new Uint8Array(length);
    let ptr = this._read(bytes.length);
    bytes.set(this.buf.subarray(ptr, ptr + length));
    return bytes;
  }
}

export let encodeUTF8: (text: string) => Uint8Array
export let decodeUTF8: (bytes: Uint8Array) => string

// For the browser and node 12.x
if (typeof TextEncoder !== 'undefined' && typeof TextDecoder !== 'undefined') {
  let encoder = new TextEncoder();
  let decoder = new TextDecoder();
  encodeUTF8 = text => encoder.encode(text);
  decodeUTF8 = bytes => decoder.decode(bytes);
}

// For node 10.x
else if (typeof Buffer !== 'undefined') {
  encodeUTF8 = text => {
    let buffer: Uint8Array = Buffer.from(text);

    // The test framework called "Jest" breaks node's Buffer API. Normally
    // instances of Buffer are also instances of Uint8Array, but not when
    // esbuild is run inside of whatever weird environment Jest uses. More
    // info: https://github.com/facebook/jest/issues/4422.
    if (!(buffer instanceof Uint8Array)) {
      // Construct a new Uint8Array with the contents of the buffer to force
      // it to be a Uint8Array instance. This is wasteful since it's slower
      // than just using the Buffer, but this should only happen when esbuild
      // is run inside of Jest.
      buffer = new Uint8Array(buffer);
    }

    return buffer;
  };
  decodeUTF8 = bytes => {
    let { buffer, byteOffset, byteLength } = bytes;
    return Buffer.from(buffer, byteOffset, byteLength).toString();
  }
}

else {
  throw new Error('No UTF-8 codec found');
}

export function readUInt32LE(buffer: Uint8Array, offset: number): number {
  return buffer[offset++] |
    (buffer[offset++] << 8) |
    (buffer[offset++] << 16) |
    (buffer[offset++] << 24);
}

function writeUInt32LE(buffer: Uint8Array, value: number, offset: number): void {
  buffer[offset++] = value;
  buffer[offset++] = value >> 8;
  buffer[offset++] = value >> 16;
  buffer[offset++] = value >> 24;
}
